timeout/timeout.1
.\"
.\" Copyright(C) 2024 Luiz Antônio Rangel. All rights reserved.
.\"
.\" SPDX-Licence-Identifier: Zlib
.\"
.TH TIMEOUT 1 "2/20/24" "Heirloom Toolchest" "User Commands"
.SH NAME
timeout \- execute a command with a time limit
.SH SYNOPSIS
\fBtimeout\fR [\fB\-k\fR \fItime\fR]
[\fB\-s\fR \fIsignal\fR] [\-fp]
\fItime\fR \fIcommand\fR [\fIargument\fR ...]
.SH DESCRIPTION
.I Timeout
executes a specified \fIcommand\fR and
attempts to terminate it after an
also specified \fItime\fR.
Per default, it sends signal 15
(terminate, alias SIGTERM), but the
signal can be redefined (see option
'\fI\-s\fR').
If the duration is a zero value, the
timer is disabled (see timer_settime(2)),
ergo there will not be any signal sent
to the specified \fIcommand\fR's process.
.PP
It accepts the following options:
.TP
.B \-k \fItime\fR
Sets a complementary time for tolerance
before sending a final signal 9 (SIGKILL)
to \fIcommand\fR if it's still executing.
.TP
.B \-s \fIsignal\fR
Sets a signal to be sent instead of
signal 15 (terminate). It's defined
per a decimal number or its symbolic
name (see kill(1) '\-l' option).
.br
If it's defined as 9 (SIGKILL), it
has an immediate effect and annuls
the '\-k' option.
.TP
.B \-f
Permits that \fIcommand\fR runs as
foreground and recieves signals
from the terminal; per using this
option, the children processes that
\fIcommand\fR may create will not be
terminated with it even in case of
the timer timing out.
.TP
.B \-p
Keep \fIcommand\fR's exit status
instead of using this program's
default.
.SH "DURATION FORMATTING"
\fItime\fR can be specified as a
fractional number by using a full
stop or a comma as the decimal
separator.
The default value are seconds
unless a time unit is specified per
its single-letter identifier; in
ascend order, the supported units
are:
.TP
.I s
seconds
.TP
.I m
minutes
.TP
.I h
hours
.TP
.I d
days
.SH DIAGNOSTICS
Exit status is 124 if the timer expired
and the \-p option was not set, 126 if
the command could be found, but could not
be executed and 127 if the command could
not be found.
.SH EXAMPLES
This program has different possible
applications in one's routine, here are some:
.br
Lets suppose you need to check if your system
is connected to the Internet and that it can
successfully connect to a server in less than
5 seconds, you can use timeout along with nc(1):
.RS 2
.sp
.nf
% timeout -p 5 nc -vz pindorama.dob.jp 80
Connection to pindorama.dob.jp (168.138.140.147) 80 port [tcp/http] succeeded!
% echo $?
0
.fi
.sp
.RE
In another example, you have a project and
start experiencing hangings at some part after
a contributor sent a patch, but you do not know
the part nor the file that was broken; since
you know that it should not take more than
half a minute to pass that part, you can use
timeout and visualize precisely where it hangs
per enabling debug at the shell interpreter:
.IP \& 2
.BI "% timeout 0,5m /bin/ksh -x ./build.ksh"
.LP
.SH NOTES
Although this implementation being \fIsœur\fR
with BSDs' implementations since it shares some
parts of code with it, it does not support long
options as GNU (and BSDs') does, so some shell
scripts may not work unchanged with this
implementation.
.br
An abideable approach for solving this problem
could be an small check/wrapper for the
timeout command in case of utilizing the \-f or
the \-p options in its long format.
.RS 2
.sp
.nf
if test \`getconf HEIRLOOM_TOOLCHEST_VERSION\` \-ge 20240220; then
     timeout \-p \-k\fI$time\fR \fI$time\fR \fI"$command"\fR
else
     timeout \-\-preserve\-duration \-\-kill\-after \fI$time\fR \fI$time\fR \fI"$command"\fR
fi
.fi
.sp
.RE
.PP
This implementation supports intervals using
both European or Anglo decimal separators,
GNU doesn't.
.PP
In contrast to OpenBSD's \(em and any other
implementation thoroughly based and/or ported
from FreeBSD 10.3R \(em, this implementation
supports nanoseconds as the interval instead
of microseconds. This can change in the future
depending on your platform, though.
.SH "SEE ALSO"
exec(3), fork(2), kill(1), signal(2),
timer_settime(2), itimerspec(3type),
wait(2)
.SH HISTORY
The timeout utility first appeared as a proposal
in BusyBox mailing lists in February 3, 2006,
but this version have payloads closer to the
timeout found at GNU Coreutils 7.0,
released in 05 October, 2008.


CHANGES
Release 250220
* seq, write: Fix memory fault due to illegal pointer access.
* cpio, pax: Fix manual page formatting.

Release 250209
* cpio: Fix manual page formatting.
* apply: Fixed the command string magic character computation algorithm that
  got garbled after a refactor made in the name of performance.
* build/mk.config, chroot, nawk: Adding -ltinfo for static compilation with
  glibc and making changes to adhere to C standards (patch by Gabriel Souza).
* [NEW] wall, write: New implementation ported from UNIX 7th (a.k.a. 32V),
  bulletproofed against buffer overflows and the CVE-2024-28085, which is
  available as the 'ssafe()' function at libcommon.
* rm: Alias '-e' flag to more known '-v' for verbose.
* date, shl: Fix compilation on platforms where (struct utmpx).ut_tv is not
  of type 'timeval'.
* rev: UTF-8 support, borrowed from suckless sbase. In varietate concordia.
* [NEW] apply, pick: Complete implementation for both apply and pick, from
  Research UNIX 8th and 10th respectively; extended payloads for pick in
  comparison with the original implementation.
* seq: Support for counting from negative numbers to positive or making a
  negative sequence.
* rev: Full refactor (made by Samuel Brederodes).
* Fixes for building dynamically on Arch Linux and other GNU C
  Library-based Linux systems which ships exclusively dynamic versions
  of its libraries, also remove `readlink_ucb' object file at `readlink/'
  build directory when executing `make clean' (patch by Molly A. McCollum).
* Remove `register's from legacy code that was living unchanged since at
  least the 80's, since modern linters started complaining about it and
  it is not used as much on most common architectures such as x86
  (patch by Samuel Brederodes).
* watch: `use_default_colors()', do not force black and white.
* [NEW] rev: New implementation, not capable of UTF-8 yet.
* cpio, diff, tabs: Fix duplicated definitions (made/reported by crumbtoo,
  merged by me, Luiz Antônio Rangel). Thanks! --- or better: Merci!
* rm: Added real support for the POSIX 1003.1(2008) standard, implemented
  IBM AIX(R)-style verbosing and 4.4BSD's '-d' option for directories ---
  rmdir(1)-esque payload --- as extensions to the standard.
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants.

Release 230503
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin.

Release 230120
* Notes for the library directories if linking against the GNU C library.

timeout/timeout.c
/*
 * timeout.c - execute a command with a time limit
 */
/*
 * Copyright (C) 2023, 2024: Luiz Antônio Rangel (takusuman)
 *                           Arthur Bacci (arthurbacci)(atr)
 *
 * SPDX-Licence-Identifier: Zlib
 *
 * The 'parse_interval' function was adapted from suspicious-tools'
 * watch(1) (available at
 * https://github.com/arthurbacci/suspicious-tools.git) and licensed
 * under the public domain.
 *
 * The part that makes the called command, in fact, sleep was
 * throloughly based on OpenBSD's implementation of timeout.
 * As per the copyright header of OpenBSD's "usr.bin/timeout/timeout.c":
 * Copyright (c) 2021 Job Snijders <job@openbsd.org>
 * Copyright (c) 2014 Baptiste Daroussin <bapt@FreeBSD.org>
 * Copyright (c) 2014 Vsevolod Stakhov <vsevolod@FreeBSD.org>
 *
 * SPDX-Licence-Identifier: BSD-2-Clause
 */

#include "sigtable.h"
#include <ctype.h>
#include <errno.h>
#include <pfmt.h>
#include <signal.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <strmenta.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>

static char *progname;

int main(int argc, char *argv[]);
int validate_signal(char *signal_name);
struct TClock validate_duration(char *duration);
int parse_interval(const char *ss, struct TClock *interval);
void settimeout(struct TClock *duration);
void handle_signal(int signo);
void usage(void);

struct TClock {
	time_t sec;
	long int nsec;
};

struct LSignal {
	sig_atomic_t sig_alrm;
	sig_atomic_t sig_chld;
	sig_atomic_t sig_term;
	sig_atomic_t sig_ign;
};
static struct LSignal siglist;

int main(int argc, char *argv[]) {
	progname = argv[0];
	register int c = 0,
		i = 0,
		s = 0;
	int option = 0,
	    fForeground = 0,
	    fOnemoretime = 0,
	    fPreserve_status = 0,
	    killer_sig = SIGTERM, /* Default killer signal is SIGTERM. */
			/*
			 * -1 can be replaced by another signal
			 * if the user sets the -s flag.
			 */
	    sigsused[] = { -1,  SIGHUP, SIGINT, SIGQUIT,
		    		SIGTERM, SIGCHLD, SIGALRM },
	    execerr = 0, /* Command execution error code. */
	    ecmd = 0, /* Command exit code. */
	    eprog = 0; /* Program exit code. */

	char **commandv,
	     *fst_commandv = NULL;
	/*
	 * atr note: it might be better to have other integers
	 * replaced by booleans, since it makes it a lot easier
	 * to tell what is being acomplished.
	 */
	bool timesout = false;
	pid_t cmdpid = 0,
	      exec_pid = 0,
	      pgid = 0;
	struct TClock first_interval = {0},
	      second_interval = {0};
	struct sigaction sa = {0};

	while ( (option = getopt(argc, argv, "fps:k:h")) != -1 ) {
		switch (option) {
			case 'f':
				fForeground = 1;
				break;
			case 'p':
				fPreserve_status = 1;
				break;
			case 's':
				killer_sig = validate_signal(optarg);
				break;
			case 'k':
				fOnemoretime = 1;
				second_interval = validate_duration(optarg);
				break;
			case 'h':
			default:
				usage();
				break;
		}
	}
	argc -= optind;
	argv += optind;

	/* Missing operand. */
	if ( argc < 2 ){
		usage();
	}

	/*
	 * atr note: if one wants to use malloc he must be sure that the last
	 * element of commandv is set to NULL
	 */
	/* Allocate commandv[], where argv[] will be copied to. */
	if ( (commandv = calloc((size_t)argc, sizeof(char *))) == NULL ) {
		pfmt(stderr, MM_ERROR, "%s: could not allocate an array of "
		                       "%d elements, each one being %lu "
				       "bytes large.\n", progname,
				       argc, sizeof(char *)
		);
		exit(1);
	}

	/* And then copy argv[] to commandv[]. */
	/* atr: note that the previous code was only copying the reference, I
	 * believe it's a mistake and replaced it by strdup.
	 */
	/* atr: it's shifted, the first element of commandv is the second of
	 * argv and so on.
	 */
	for (c = 1; c < argc; c++) {
		commandv[c - 1] = strdup(argv[c]);
	}

	/* atr: dummy string for funny mutations */
	fst_commandv = strdup(argv[0]);
	first_interval = validate_duration(fst_commandv);
	free(fst_commandv);

	if (! fForeground) {
		/*
		 * setpgid(pid, pgid)
		 * If both the 'pid' and 'pgid' are equal to zero, setpgid(2)
		 * assumes that both the child process pid and pgid shall be the
		 * same as parent's, that is why it is used in this case, where
		 * we do not want to foreground the process.
		 */
		if ((pgid = setpgid(0, 0)) != 0) {
			pfmt(stderr, MM_ERROR,
			     "%s: failed to set process group via setpgid"
			     "(0, 0): %s.\n", progname, strerror(errno));
		}
	}

	/*
	 * The list of signals masked (blocked) during the execution of the
	 * handler (which is present at sigaction.sa_handler/sa_sigaction)
	 * must be initialized per excluding the current list at sa_mask.
	 */
	sigemptyset(&sa.sa_mask);

	/*
	 * Since these two signals can not be caught nor ignored (using
	 * sa_mask), its inclusion at the signal ignore array must avoided.
	 * It will not cause any error, though, just be ignored.
	 */
	if (killer_sig != SIGKILL && killer_sig != SIGSTOP) {
		sigsused[0] = killer_sig;
	}

	/*
	 * Iterate over the sigsused[] integer array and start adding
	 * each one to the sa_mask set again.
	 */
	for (s = 0; s < (int)(sizeof(sigsused)/sizeof(sigsused[0])); s++) {
		sigaddset(&sa.sa_mask, sigsused[s]);
	}

	sa.sa_handler = &handle_signal;

	/*
	 * Set SA_RESTART at sa_flags, this will make certain system calls
	 * restartable by signals. This is useful because, after handle_signal()
	 * returns, the system call is restarted instead of failing with EINTR,
	 * which means an error by interruption.
	 */
	sa.sa_flags = SA_RESTART;

	/*
	 * Iterate over the sigsused[] integer array again, but this time it
	 * will be using sigaction() for changing an action taken by the system
	 * when any of the signals of the array --- which also are present on the
	 * handle_signal() function and, as the name explicitaly say, handled by
	 * it --- are sent, so, when any of these signals are sent, it actually
	 * calls handle_signal().
	 */
	for (s = 0; s < (int)(sizeof(sigsused)/sizeof(sigsused[0])); s++) {
		if (sigsused[s] != -1 && sigsused[s] != 0
		&& sigaction(sigsused[s], &sa, NULL) != 0) {
			pfmt(stderr, MM_ERROR,
				"%s: failed to change action for %s: %s.\n",
            			progname, strsignal(sigsused[s]),
					strerror(errno));
			exit(1);
		}
	}

	/*
	 * This is meant to ignore stop signals that could come from the
	 * terminal on the parent process.
	 */
	signal(SIGTTIN, SIG_IGN);
	signal(SIGTTOU, SIG_IGN);

	exec_pid = fork();

	/* Get the bad news first. */
	if (exec_pid == -1) {
		pfmt(stderr, MM_ERROR, "%s: failed to fork: %s.\n",
        	    progname, strerror(errno));
	} else if (exec_pid == 0) {
		/*
		 * On the child process, signals sent by the terminal mustn't
		 * be ignored anymore and that the default action --- without
		 * involving the handler --- have to be taken by the system.
		 */
		signal(SIGTTIN, SIG_DFL);
		signal(SIGTTOU, SIG_DFL);

		/* Here we will be executing the child process. */
		if ( (execvp(commandv[0], commandv)) == -1 ) {
			pfmt(stderr, MM_ERROR, "%s: failed to exec(): %s.\n",
					progname, strerror(errno));
		}

		/*
		 * According to GNU's timeout(1) manual page:
		 *
		 *  126    if COMMAND is found but cannot be invoked
		 *  127    if COMMAND cannot be found
		 *
		 *  ENOENT stands for "error no entity/entry", so it
		 *  means, in this context, no such file or directory.
		 */
		execerr = (errno == ENOENT) ? 127 : 126;
		_exit(execerr);
	}

	/* The command string will not be necessary after exec(). */
	for (i = 0; commandv[i]; i++) {
		/* atr: frees the strduped strings */
		free(commandv[i]);
	}
	free(commandv);

	/*
	 * Add signals from the sa_mask set to the SIG_BLOCK set,
	 * blocking it after the execution of the commmand.
	 */
	if (sigprocmask(SIG_BLOCK, &sa.sa_mask, NULL) != 0) {
		pfmt(stderr, MM_ERROR, "%s: failed to sigprocmask(): %s.\n",
				progname, strerror(errno));
		exit(1);
	}

	settimeout(&first_interval);

	for (;;) {
		/*
		 * The sa_mask will be empty again, but we will
		 * also be using sigsuspend() to prevent the
		 * function assigned to sa_handler of
		 * interrupting this section.
		 */
		sigemptyset(&sa.sa_mask);
		sigsuspend(&sa.sa_mask);

		if (siglist.sig_chld) {
			siglist.sig_chld = 0;

			/*
			 * taks note: This code looks a little confusing for
			 * people who never worked with signals in C, so there's
			 * my explanation:
			 * If the signal sent was a SIGCHLD, keep an eye if the
			 * wait(2) function returns that the child process had
			 * been interrupt (EINTR), in that case, continue the
			 * code, making the program exit code (eprog) equal to
			 * the command exit code (ecmd) and getting loose from
			 * the signal-checking loop.
			 */
			for (; ((cmdpid = wait(&ecmd)) < 0 && errno == EINTR);) {
				continue;
			}

			if (cmdpid == exec_pid) {
				eprog = ecmd;
				break;
			}
		} else if (siglist.sig_alrm || siglist.sig_term) {
			if (siglist.sig_alrm) {
				siglist.sig_alrm = 0;
				timesout = true;
			}

			if (! fForeground) {
				killpg(pgid, killer_sig);
			} else {
				kill(exec_pid, siglist.sig_term
						? (int)siglist.sig_term
						: killer_sig);
			}

			if (! fOnemoretime) {
				break;
			} else {
				settimeout(&second_interval);
				fOnemoretime = 0;
				siglist.sig_ign = killer_sig;
				killer_sig = SIGKILL;
			}
		}
	}

	for (; (cmdpid != exec_pid && wait(&eprog) == -1);) {
		if (errno != EINTR) {
			pfmt(stderr, MM_ERROR, "%s: failed to wait(): %s.\n",
			progname, strerror(errno));

			exit(1);
		}
	}

	if (WEXITSTATUS(eprog)) {
		eprog = WEXITSTATUS(eprog);
	} else if (WIFSIGNALED(eprog)) {
		eprog = (128 + WTERMSIG(eprog));
	}

	if (timesout && !fPreserve_status) {
		eprog = 124;
	}

	return eprog;
}

void settimeout(struct TClock *duration) {
	timer_t tid;
	struct itimerspec its;

	/*
	 * Since the second value, sevp on the prototype of
	 * timer_create(2), is defined as NULL, every
	 * expiration of this timer will sent an SIGALRM.
	 * Just as expected.
	 */
	if (timer_create(CLOCK_REALTIME, NULL, &tid) != 0) {
		pfmt(stderr, MM_ERROR,
			"%s: failed to create timer: %s\n",
			progname, strerror(errno));
		exit(1);
	}

	/* Timer expiration. */
	its.it_value.tv_sec = duration->sec;
	its.it_value.tv_nsec = duration->nsec;
	/*
	 * Timer period betwixt expirations.
	 * taks note: This must be all zeroed,
	 * ergo option '-k' can work properly.
	 * itimerspec(3type) is incomplete
	 * on Linux and reads vague on NetBSD,
	 * so I'm doing pure Computer Alchemy here.
	 */
	its.it_interval.tv_sec = 0;
	its.it_interval.tv_nsec = 0;

	if (timer_settime(tid, 0, &its, NULL) != 0) {
		pfmt(stderr, MM_ERROR,
			"%s: failed to set timer: %s\n",
			progname, strerror(errno));
		exit(1);
	}
}

int validate_signal(char *str) {
	register int i;

	/*
	 * Check if the first character of the input
	 * string is a letter, so it can be parsed
	 * into a signal name.
	 */
	if (isalpha(str[0])) {
		/* If it starts with "SIG" */
		if (strncmp(str, "SIG", 3) == 0) {
			/* Skip 3 bytes (SIG) */
			str = &str[3];
		}

		for (i = 0; sig_strs[i].signame; i++) {
			if (strcmp(str, sig_strs[i].signame) == 0)
				return sig_strs[i].signum;
		}

		pfmt(stderr, MM_ERROR, "%s: invalid signal %s.\n",
					progname,          str);
		exit(1);
	} else {
		i = atoi(str);

		if (i < 0 || i > SIGRTMAX) {
			pfmt(stderr, MM_ERROR, "%s: invalid signal %d.\n",
			                        progname,          i);
			exit(1);
		}
		return i;
	}
}

/* Boilerplate for parse_interval(). */
struct TClock validate_duration(char *timestr) {
	struct TClock mclock;

	if (parse_interval(timestr, &mclock) == -1) {
		pfmt(stderr, MM_ERROR,
			"the given interval couldn't be parsed.\n");
		exit(1);
	}

	return mclock;
}

/*
 * Returns -1 if the given string is invalid,
 * otherwise 0 is returned and the interval
 * struct is set.
 */
int parse_interval(const char *ss, struct TClock *interval) {
	char s[32] = "",
	     *afterpoint = NULL,
	     *decsep = NULL,
	     *tunit = NULL;
	/* 'double' for the converted time per unit. */
	double ftime = 0;
	size_t i = 0,
	       afterpoint_len = 0;

	/*
	 * Support both Anglo and European decimal separators.
	 * taks note: In varietate concordia. 🇧🇷🇪🇺🤝🇺🇸🇬🇧
	 */
	decsep = strchr(ss, ',');
	if (decsep) {
		*decsep = '.';
	}

	ftime = strtod(ss, &tunit);
	switch (tunit[0]) {
		case '\0':
		case 's':
			break;
		case 'm':
			ftime *= 60;
			break;
		case 'h':
			ftime *= (60 * 60);
			break;
		case 'd':
			ftime *= ((60*60) * 24);
			break;
		default:
			pfmt(stderr, MM_ERROR,
				"%s: invalid time unit suffix '%c'.\n",
					progname, tunit);
			exit(1);
	}

	/* Copy the time value to 's'. */
	snprintf(s, 32, "%g", ftime);

	for (i = 0; s[i]; i++) {
		switch (isdigit(s[i])) {
			case 0:
				switch (s[i]) {
					case '.':
						continue;
					default:
						break;
				}
				return -1;
			default:
				continue;
		}
	}

	i = afterchar(s, '.');
	if (afterpoint == NULL) {
		if (i != 0 && s[0] != '.') {
			afterpoint = &s[i + 1];
			s[i] = '\0';
		}
	} else {
		return -1;
	}

	interval->sec = strlen(s)
			? atoi(s)
			: 0;
	interval->nsec = afterpoint
			? atoi(afterpoint)
			: 0;

	if (afterpoint == NULL) {
	       	return 0;
	}

	afterpoint_len = strlen(afterpoint);
	if (afterpoint_len > 9) {
	       	return -1;
	}
	for (i = afterpoint_len; i < 9; i++) {
		interval->nsec *= 10;
	}

	return 0;
}

/*
 * This is just a function that sets variables --- which
 * indicates which signals where enabled or not ---, so it
 * does not return anything. I wish these could be returned
 * on something like a struct, like it was done for the
 * interval, but the UNIX ABI does not give this opportunity.
 */
void handle_signal(int signo) {
	if (siglist.sig_ign != 0 && signo == siglist.sig_ign) {
		/* Ignore this signal. */
		siglist.sig_ign = 0;
		return;
	}

	switch (signo) {
		case SIGHUP:
		case SIGINT:
		case SIGQUIT:
		case SIGTERM:
			siglist.sig_term = signo;
			break;
		case SIGCHLD:
			siglist.sig_chld = 1;
			break;
		case SIGALRM:
			siglist.sig_alrm = 1;
			break;
	}
}

void usage(void) {
	pfmt(stderr, MM_NOSTD, "usage: %s: [-fp] [-s signal] [-k time] "
	                       "time command [args ...]\n", progname);
	exit(1);
}

